overlord/servicestate/quota_control.go

```diff
@@ -68,8 +102,136 @@ func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[s
+	grpsToStart := []*quota.Group{}
+	appsToRestartBySnap := map[*snap.Info][]*snap.AppInfo{}
+
+	collectModifiedUnits := func(app *snap.AppInfo, grp *quota.Group, unitType string, name, old, new string) {
+		switch unitType {
+		case "slice":
+			// this slice was either modified or written for the first time
+
+			// There are currently 3 possible cases that have different
+			// operations required, but we ignore one of them, so there really
+			// are just 2 cases we care about:
+			// 1. If this slice was initially written, we just need to systemctl
+			//    start it
+			// 2. If the slice was modified to be given more resources (i.e. a
+			//    higher memory limit), then we just need to do a daemon-reload
+			//    which causes systemd to modify the cgroup which will always
+			//    work since a cgroup can be atomically given more resources
+			//    without issue since the cgroup can't be using more than the
+			//    current limit.
+			// 3. If the slice was modified to be given _less_ resources (i.e. a
+			//    lower memory limit), then we need to stop the services before
+			//    issuing the daemon-reload to systemd, then do the
+			//    daemon-reload which will succeed in modifying the cgroup, then
+			//    start the services we stopped back up again. This is because
+			//    otherwise if the services are currently running and using more
+			//    resources than they would be allowed after the modification is
+			//    applied by systemd to the cgroup, the kernel responds with
+			//    EBUSY, and it isn't clear if the modification is then properly
+			//    in place or not.
+			//
+			// We will already have called daemon-reload at the end of
+			// EnsureSnapServices directly, so handling case 3 is difficult, and
+			// for now we disallow making this sort of change to a quota group,
+			// that logic is handled at a higher level than this function.
+			// Thus the only decision we really have to make is if the slice was
+			// newly written or not, and if it was save it for later
+			if old == "" {
+				grpsToStart = append(grpsToStart, grp)
+			}
+
+		case "service":
+			// in this case, the only way that a service could have been changed
+			// was if it was moved into or out of a slice, in both cases we need
+			// to restart the service
+			sn := app.Snap
+			appsToRestartBySnap[sn] = append(appsToRestartBySnap[sn], app)
+
+			// TODO: what about sockets and timers? activation units just start
+			// the full unit, so as long as the full unit is restarted we should
+			// be okay?
+		}
+	}
+	if err := wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, collectModifiedUnits, progress.Null); err != nil {
+		return err
+	}
+
+	// now start the slices
+
+	// TODO: should this logic move to wrappers in wrappers.RestartGroups() ?
+	systemSysd := systemd.New(systemd.SystemMode, progress.Null)
+
+	for _, grp := range grpsToStart {
+		// TODO: what should these timeouts for stopping/restart slices be?
+		if !ensureOpts.Preseeding {
```

Comment:
we can extract and move this kind of check at 162 because

Reply:
I think see what you mean, but for what it's worth this comment is a bit difficult to parse :smile: